### ISSUE TITLE
comments not assigned to `If` nodes when using a binary operator

### ISSUE THREAD

**Issue by etamponi** [jashkenas/coffeescript#4781]
```
$ coffee -v
CoffeeScript version 2.0.2
```

I am trying to port coffee-coverage to CoffeeScript 2, I've got all the tests passing, but I am having some issues with pragma comments.

I would expect this code:

```
### istanbul ignore if ###
if something is "abc"
  console.log "Hello"
```

To generate an AST in which the `If` node has a `comments` field containing `[ " istanbul ignore if " ]`. Instead, the comments are assigned to first element of the binary operation, so to the `something` variable.

I think this is a bug, since this code, instead, works as expected:

```
### istanbul ignore if ###
if something
  console.log "Hello"
```

the comments field is assigned correctly to the `If` node and its children.

**Comment by GeoffreyBooth** [comment]
So the way to troubleshoot comments is to save your test code in a file, e.g. `test.coffee`, and compile like this:

```bash
DEBUG_TOKEN_STREAM=1 DEBUG_REWRITTEN_TOKEN_STREAM=1 coffee -bp ./test.coffee
```

In the case of your first example, I get this:

```
Initial token stream:
TERMINATOR/
 JS/* TERMINATOR/
 IF/if IDENTIFIER/something COMPARE/== STRING/"abc" INDENT/2 IDENTIFIER/console ./. PROPERTY/log STRING/"Hello" OUTDENT/2 TERMINATOR/

Rewritten token stream:
JS/* TERMINATOR/
 IF/if IDENTIFIER/something COMPARE/== STRING/"abc" INDENT/2 IDENTIFIER/console ./. PROPERTY/log CALL_START/( STRING/"Hello" CALL_END/) OUTDENT/2 TERMINATOR/

/* istanbul ignore if */
if (something === "abc") {
  console.log("Hello");
}
```

Generally, comments are attached to the _preceding_ token. If a file starts with a comment, a zero-length `PassthroughLiteral` node is created (i.e. ``` `` ```) and the comment is attached to that. That’s what you see here.

If the preceding token doesn’t survive parsing, for example if it’s a closing `)`, the comment instead gets attached to the next closest token. Also if a comment is on a line by itself, it often gets moved onto the following token. See [rewriter.coffee](http://coffeescript.org/annotated-source/rewriter.html).

So in other words, you shouldn’t treat comment attachments as something predictable. A comment could appear inline anywhere, for example `if ### comment ### something`. In this case would you expect it attached to `if` or to `something`? Comments aren’t nodes like other code are. (`###` comments _were_ nodes in CoffeeScript 1, but in CoffeeScript 2 they became this “token attachment” thing instead, so that `###` comments could appear anywhere and not just where expressions were allowed.)

Sorry, I feel like this answer probably isn’t very helpful, but I’m not sure what I would recommend to help you with your use case.

**Comment by etamponi** [comment]
Well, I just tried to port the coffee-coverage to CoffeeScript 2. In CoffeeScript 1, comments get their own node in the AST, so it's easy to then find the closest non-comment node.

I managed to make coffee-coverage work with CoffeeScript 2 too though, but I found this strange behavior (that your comment doesn't explain, btw), so I had to work around it. I thought it was a bug, glad to know it is not :)

Thank you for your time!

**Comment by GeoffreyBooth** [comment]
Comments aren’t nodes anymore because if they were, they would have to follow grammar rules for where they could appear. You wouldn’t be able to write `1 + ### comment ### 2`, for example, like you can’t in CoffeeScript 1. We want the ability to write comments like that so that we can use the Flow comments-based syntax, to name one use case.

**Comment by InPermutation** [comment]
> If the preceding token doesn’t survive parsing, for example if it’s a closing `)`, the comment instead gets attached to the next closest token.

This makes sense to me, but doesn't appear to be what is happening when there is a binary operation in an `if` condition. What is actually observed in this case is that the comment skips the `if` token and is instead attached to the next token. If there is no binary operation, the comment is attached to the `if` as expected.

**Comment by GeoffreyBooth** [comment]
`'IF'` is one of the discarded tokens. Look at the bottom of `rewriter.coffee`:

```coffee
DISCARDED = ['(', ')', '[', ']', '{', '}', '.', '..', '...', ',', '=', '++', '--', '?',
  'AS', 'AWAIT', 'CALL_START', 'CALL_END', 'DEFAULT', 'ELSE', 'EXTENDS', 'EXPORT',
  'FORIN', 'FOROF', 'FORFROM', 'IMPORT', 'INDENT', 'INDEX_SOAK', 'LEADING_WHEN',
  'OUTDENT', 'PARAM_END', 'REGEX_START', 'REGEX_END', 'RETURN', 'STRING_END', 'THROW',
  'UNARY', 'YIELD'
].concat IMPLICIT_UNSPACED_CALL.concat IMPLICIT_END.concat CALL_CLOSERS.concat CONTROL_IN_IMPLICIT
```

`'IF'` is inside `CONTROL_IN_IMPLICIT`. A comment can end up attached to an `If` _node_ because it gets “bubbled up” from one of the node’s children. If you look in `nodes.coffee` there are lots of cases where comments get moved around between nodes because some nodes get discarded during compilation.
